### PR TITLE
Add hydra_endpoints relation interface

### PR DESCRIPTION
**Pull request by natalian98**
### Description
Kratos-operator requires `oauth2_provider` parameter in order to be able to interact with hydra-operator. This parameter is the Hydra's admin API endpoint, which will usually be exposed through Ingress using `traefik-k8s` charm. We aim to eliminate the need of manually passing config values to each charm and have them related instead.
To summarize this relation, Hydra needs to provide its endpoints to Kratos.

### Future work
The interface may be subject to change in case the future upstream releases of kratos and/or hydra require different parameters in order to integrate with each other.

### CODE REVIEW

File: interfaces/hydra_endpoints/v0/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# `hydra_endpoints`
+
+## Usage
+
+This relation interface describes the expected behavior of provider ([Ory Hydra](https://github.com/canonical/hydra-operator))
+and requirer ([Ory Kratos](https://github.com/canonical/kratos-operator)) sides of relation.
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:6]: Here I'd use the same as other relation interfaces: `This relation interface describes the expected behavior of charms claiming to be able to provide or consume an hydra endpoint`

File: interfaces/hydra_endpoints/v0/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+Kratos requires `oauth2_provider` parameter in order to be able to interact with Hydra. This parameter is the Hydra's admin API endpoint,
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:8]: This explanation is unnecessary here but can be in the PR description to add context

File: interfaces/hydra_endpoints/v0/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+which will usually be exposed through Ingress using `traefik-k8s` charm.
+The interface will provide admin and public endpoints, in case any other charm belonging to the IAM bundle requires them both.
+
+The purpose of this relation is to facilitate integration between the two charms by passing the endpoints from Hydra to Kratos.
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:12]: This sentence seems also unnecessary here

File: interfaces/hydra_endpoints/v0/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+For now, the only known requirer is Kratos and Hydra is the only provider, however the interface may be reused by other charms,
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:14]: Again, not needed here

File: interfaces/hydra_endpoints/v0/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+as long as they provide or require admin and/or public endpoints.
+
+## Direction
+
+The interface will consist of a provider and a requirer. The provider is expected to supply its public and admin endpoints,
+while the requirer will just consume the information.
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:20]: `just consume` is too abstract

File: interfaces/hydra_endpoints/v0/README.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+```mermaid
+flowchart
+    Requirer ----> Provider
+    Provider -- hydra_admin_endpoint, hydra_public_endpoint --> Requirer
+```
+
+## Behavior
+
+The provider must be able to supply admin and public endpoints. There are no criteria for the requirer as it's supposed to just consume the relation data.
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:30]: I'd use the generic `Both the Requirer and the provider need to adhere to a certain set of criterias to be considered compatible with the interface.` and add details in the subbullets

File: interfaces/hydra_endpoints/v0/README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+### Provider
+
+- Is expected to send the public and admin URLs to the interface.
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:34]: `send to the interface` is not really precise, `write to the application relation databag` seems more accurate

File: interfaces/hydra_endpoints/v0/README.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  "public_endpoint": "public-endpoint"
+}
+}
+```
+
+### Requirer
+
+[\[JSON Schema\]](./schemas/requirer.json)
+
+#### Example
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:67]: There shouldn't be anything here as far as I understood the rest of the interface.

File: interfaces/hydra_endpoints/v0/schemas/requirer.json [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    "$id": "https://canonical.github.io/charm-relation-interfaces/interfaces/hydra_endpoints/schemas/requirer.json",
+    "title": "`hydra_endpoints` requirer schema",
+    "description": "The `hydra_endpoints` root schema comprises the entire requirer databag for this interface.",
+    "default": {},
+    "type": "object",
+    "properties": {
+        "application_data": {
+            "type": "object",
+            "properties": {
+                "admin_endpoint": {
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/schemas/requirer.json:12]: Those shouldn't be here, the requirer isn't writing data to the databag.

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: interfaces/hydra_endpoints/v0/README.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+## Behavior
+
+The provider must be able to supply admin and public endpoints. There are no criteria for the requirer as it's supposed to just consume the relation data.
+
+### Provider
+
+- Is expected to send the public and admin URLs to the interface.
+  The url is expected to have the following structure:
+
+    > `http://[traefik-hostname]:[traefik-port]/[app-name]-[model-name]/`
```
Comment (nsklikas) [inline interfaces/hydra_endpoints/v0/README.md:37]: The URL is not required to have this structure, this is true only if the user is using charmed traefik to access Hydra

**Comment by jnsgruk** [comment]
This needs adding to the top-level README, but otherwise is starting to look good!

File: interfaces/hydra_endpoints/v0/README.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+#### Example
+
+
+```json
+{
+"application_data": {
+  "admin_endpoint": "admin-endpoint",
+  "public_endpoint": "public-endpoint"
+}
+}
```
Comment (jnsgruk) [inline interfaces/hydra_endpoints/v0/README.md:50]: Nit: something isn't right with the indentation here

File: interfaces/hydra_endpoints/v0/README.md [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+- The provider must be able to supply admin and public endpoints
+- There are no criteria for the requirer as it's supposed to just consume the relation data.
+
+### Provider
+
+- Is expected to write the public and admin URLs to the application databag.
+
+### Requirer
+
+- Is expected to consume the relation data.
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:32]: This feels minimal, what should it do with those URL's

File: interfaces/hydra_endpoints/v0/README.md [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+## Behavior
+
+Both the requirer and the provider need to adhere to a certain set of criteria to be considered compatible with the interface:
+- The provider must be able to supply admin and public endpoints
+- There are no criteria for the requirer as it's supposed to just consume the relation data.
+
+### Provider
+
+- Is expected to write the public and admin URLs to the application databag.
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:28]: This feels minimal, the provider should probably serve something?

File: interfaces/hydra_endpoints/v0/README.md [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+flowchart
+    Requirer ----> Provider
+    Provider -- hydra_admin_endpoint, hydra_public_endpoint --> Requirer
+```
+
+## Behavior
+
+Both the requirer and the provider need to adhere to a certain set of criteria to be considered compatible with the interface:
+- The provider must be able to supply admin and public endpoints
+- There are no criteria for the requirer as it's supposed to just consume the relation data.
```
Comment (gruyaume) [inline interfaces/hydra_endpoints/v0/README.md:24]: those 2 lines aren't necessary as the subbullets below will describe those same behaviors

File: interfaces/hydra_endpoints/v0/README.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+Both the requirer and the provider need to adhere to a certain set of criteria to be considered compatible with the interface:
+
+### Provider
+
+- Is expected to serve admin and public API endpoints 
+- Is expected to write the public and admin URLs to the application databag.
+
+### Requirer
+
+- Is expected to consume the relation data by updating its config file with the received endpoints.
```
Comment (simskij) [inline interfaces/hydra_endpoints/v0/README.md:31]:
```suggestion
- Is expected to consume the relation data to set up integration with Hydra.
```

File: interfaces/hydra_endpoints/v0/README.md [hunk 16]
```diff
@@ -0,0 +1,56 @@
+# `hydra_endpoints`
+
+## Usage
```
Comment (simskij) [inline interfaces/hydra_endpoints/v0/README.md:3]: Please add a one-liner `## Overview` section above this one.

**Review by simskij — CHANGES_REQUESTED** [review]

**Review by gruyaume — APPROVED** [review]

**Review by simskij — APPROVED** [review]

**Comment by jnsgruk** [comment]
thanks @natalian98!

**Pull request merged**